644 Reduce times session is duplicated as a hash in check answers controller

reggieb opened the pull request:
What
----
An Intermittent bug appears to be caused by either 

- errors in the [**json-schema** gem](https://github.com/ruby-json-schema/json-schema). The main JSON::Validator class has lots of class variables, deeply nested methods (e.g the build_schemas method is 65 lines), and a lot of the processing is done at the class level - so against the class singleton (34 methods within the class << self declaration). I expect that's why they're using Mutex to try and tame threading issues. This PR replaces json-schema with the [**json_schemer** gem](https://github.com/davishmcclurg/json_schemer) which has a simpler structure and less reliance of processing within class singletons.
- missing keys in a hash (derived from session) on being passed to JSON schema validator. The changes in this PR reduce the manipulation of session in case this is the cause of the error.

The bug generates errors like: 
```
[ The property '#/' did not contain a required property of 'name' in schema b48faf71-f8f7-5ab7-9536-fbf40922a7fa , The property '#/' did not contain a required property of 'know_nhs_number' in schema b48faf71-f8f7-5ab7-9536-fbf40922a7fa ]
```
This error can be replicated if keys are removed from the hash passed to the JSON validator.

Prior to this change, `session` was being manipulated by being converted into a hash and then converted to a `ActiveSupport::HashWithIndifferentAccess` (I believe each change creates a duplicate object in memory). As `session` already has indifferent access (keys can be identified by symbol or string versions of the keys) these changes are unnecessary manipulation.

The one place where manipulation is needed is when `except` needs to be called on the session hash (which doesn't respond to `except`). Here, as long a string keys are used in the `except` clause, just a single `to_h` manipulation is needed.

Links
-----

- [Trello card](https://trello.com/c/5BDly30W/644-investigate-formresponseinvaliderror-in-sentry)
- [Errors in sentry](https://sentry.io/organizations/govuk/issues/1667664108/?project=5170680)

reggieb committed:
644 Reduce times session is duplicated as a hash
and then as a ActiveSupport::HashWithIndifferentAccess

Intermittent bug appears to be caused by keys being missing in hash
derived from session passed to JSON schema validator. Reducing
manipulation of session in case this is cause.

steveholmes reviewed (approved):
Yes, hard to tell what will fix the problem as we can't reproduce locally, but looks good to me :+1:

reggieb committed:
Change json validation gem as issues look likely to be from original gem

The json-schema gem  The main JSON::Validator class has lots of class
variables, deeply nested methods, and a lot of the processing is done
at the class level (so against the class singleton), which can lead to
threading issues.

The json_schema gem is simpler and most processing is done at the instance
level.
https://github.com/davishmcclurg/json_schemer

Simplified schema helper specs so they just test added functionality
within app rather than behaviour of gem or validity of schema

leenagupte commented on Gemfile:
```diff
@@ -14,7 +14,7 @@ gem "faraday", "~> 1.0.1"
 gem "govuk_app_config", "~> 2.2.0"
 gem "govuk_publishing_components", "~> 21.57.1"
-gem "json-schema", "~> 2.8.1"
+gem "json_schemer", "~> 0.2"
```
@reggieb Can you explain why you're using this gem in the commit message. i.e. What's the benefit of this change? Also, can you add a link to the gem documentation.

reggieb replied:
I've updated the commit message

leenagupte commented on spec/helpers/schema_helper_spec.rb:
```diff
@@ -4,332 +4,26 @@
-          "session_id",
-        )
+      it "returns a fail as true" do
+        expect(subject.fail).to be(true)
```
I don't understand what this is testing. What does "fail" mean?

reggieb replied:
I'll rename it `invalid`

leenagupte replied:
I'm still not sure what this is testing.

Could you also add a test for if there are no validation errors? For example, what does `validate_against_form_response_schema` return if the data validates against the schema?

leenagupte reviewed (dismissed):
@reggieb I'm concerned by the changes made in the second commit. Most especially the removal of the tests used to make sure that the schema validates as expected. 

None of us are experts at writing schemas, so we used these tests to make sure that we are writing the schema correctly.

For example, we used the tests to make sure that we aren't doing things like making a value required when it isn't supposed to be, or that we aren't allowing users to just save any value in a field.

Also, we had plans to enforce the json validation at the model level, and these tests were added to give us confidence that we wouldn't be preventing users from submitting the form because the schema wasn't quite right. The model validation was delayed due to the occurrences of the error you're trying to fix in the first commit, but it's possible that it _may_ be added sometime in the future.

Also, as mentioned in the inline comments, can you explain why the json_schemer is being used. What is the benefit of it over using the existing json-schema gem?

Happy to talk this through in person.

leenagupte commented on spec/helpers/schema_helper_spec.rb:
```diff
@@ -4,332 +4,26 @@
-        data = valid_data.merge(medical_conditions: "Foo")
-        expect(validate_against_form_response_schema(data).first).to include("medical_conditions")
-      end
+    it "returns no error if the data is valid" do
```
```suggestion
    it "returns no errors if the data is valid" do
```

leenagupte commented on spec/helpers/schema_helper_spec.rb:
```diff
@@ -4,332 +4,26 @@
-      it "returns a list of errors when check_answers_seen has an invalid value" do
-        data = valid_data.merge(check_answers_seen: "Foo")
-        expect(validate_against_form_response_schema(data).first).to include("check_answers_seen")
+      it "returns a list of errors when live_in_england is missing" do
```
It looks all of the schema validation tests have been moved into a service in the next commit. Should this test join them there?

leenagupte commented on app/controllers/coronavirus_form/check_answers_controller.rb:
```diff
@@ -16,7 +16,7 @@ def submit
 
     json_validation = validate_against_form_response_schema(sanitised_session)
-    if json_validation.fail
+    if json_validation.invalid?
```
```suggestion
    unless json_validation.valid?
```
Generally `valid?` is called to test if something is ok. It would be nice if we could repeat that pattern here.

Can this change in name be fixed up to the previous commit to make it easier to traverse the history.

reggieb replied:
I've tended to find people don't like `unless` so used a form that matched `if` cleanly. Happy to go for `unless` pattern.

leenagupte commented on app/helpers/schema_helper.rb:
```diff
@@ -2,14 +2,7 @@ module SchemaHelper
-      fail: errors.any?,
-      errors: errors.map { |report| report["details"] },
-    )
+    JsonValidator.new schema, data
```
```suggestion
    JsonValidator.new(schema, data)
```

This is much nicer to read!

leenagupte commented on app/services/json_validator.rb:
```diff
@@ -0,0 +1,38 @@
+
+  def invalid?
+    !validator.valid?(data)
+  end
```
```suggestion
  def valid?
    validator.valid?(data)
  end
```

I find truthy statements are easier to read.

leenagupte commented on spec/services/json_validator_spec.rb:
```diff
@@ -0,0 +1,517 @@
+      expect(subject.invalid?).to be(false)
+    end
+
+    context "with invalid data" do
+      let(:data) { valid_data.except(:live_in_england) }
+
+      it "returns true" do
+        expect(subject.invalid?).to be(true)
+      end
```
This looks like it's a duplicate of the "live in england" test below, so can be removed.

reggieb replied:
It's testing a different method. `valid` and `errors` are two different methods, using two different calls on the gem. `valid` returns a simple boolean, where as `validate` returns an errors hash. So yes it's the same invalid data structure - it's a known invalid data structure.

leenagupte commented on spec/services/json_validator_spec.rb:
```diff
@@ -0,0 +1,517 @@
+  describe "#invalid?" do
+    it "returns false if data is valid" do
+      expect(subject.invalid?).to be(false)
+    end
```
As stated before, I think this would be easier to read if the method was testing that the data is valid rather than invalid.

leenagupte commented on spec/services/json_validator_spec.rb:
```diff
@@ -0,0 +1,517 @@
+      end
+    end
+
+    context "when live_in_england in valid" do
```
```suggestion
    context "when live_in_england is invalid" do
```

leenagupte commented on spec/services/json_validator_spec.rb:
```diff
@@ -0,0 +1,517 @@
+    context "when live_in_england in valid" do
+      let(:data) { valid_data.merge(live_in_england: "invalid") }
+
+      it "returns a yes_no definition error against field" do
```
```suggestion
      it "returns a yes_no definition error against the field" do
```

leenagupte commented on spec/services/json_validator_spec.rb:
```diff
@@ -0,0 +1,517 @@
+      expect(subject.errors).to be_empty
+    end
+
+    context "when live_in_england missing" do
```
```suggestion
    context "when live_in_england is missing" do
```

leenagupte commented on spec/services/json_validator_spec.rb:
```diff
@@ -0,0 +1,517 @@
+      end
+    end
+
+    context "when nhs_letter missing" do
```
```suggestion
    context "when nhs_letter is missing" do
```

leenagupte commented on spec/services/json_validator_spec.rb:
```diff
@@ -0,0 +1,517 @@
+      end
+    end
+
+    context "when nhs_letter invalid" do
```
```suggestion
    context "when nhs_letter is invalid" do
```

leenagupte commented on spec/services/json_validator_spec.rb:
```diff
@@ -0,0 +1,517 @@
+    context "when nhs_letter invalid" do
+      let(:data) { valid_data.merge(nhs_letter: "invalid") }
+
+      it "returns a nhs_letter property error" do
```
```suggestion
      it "returns an nhs_letter property error" do
```

leenagupte commented on spec/services/json_validator_spec.rb:
```diff
@@ -0,0 +1,517 @@
+      end
+    end
+
+    context "when medical_conditions blank" do
```
```suggestion
    context "when medical_conditions is blank" do
```

leenagupte commented on spec/services/json_validator_spec.rb:
```diff
@@ -0,0 +1,517 @@
+    context "when medical_conditions blank" do
+      let(:data) { valid_data.except(:medical_conditions) }
+
+      it "returns no error" do
```
```suggestion
      it "returns no errors" do
```

leenagupte commented on spec/services/json_validator_spec.rb:
```diff
@@ -0,0 +1,517 @@
+      end
+    end
+
+    context "when medical_conditions have unexpected value" do
```
```suggestion
    context "when medical_conditions has an unexpected value" do
```

leenagupte commented on spec/services/json_validator_spec.rb:
```diff
@@ -0,0 +1,517 @@
+      end
+    end
+
+    context "when name missing" do
```
```suggestion
    context "when name is missing" do
```

leenagupte commented on spec/services/json_validator_spec.rb:
```diff
@@ -0,0 +1,517 @@
+      end
+    end
+
+    context "within name" do
```
```suggestion
    context "#name" do
```

reggieb replied:
Name isn't an instance method, it's a key within a hash. So `#name` is incorrect. I'll change it to `'within "name"'`

leenagupte commented on spec/services/json_validator_spec.rb:
```diff
@@ -0,0 +1,517 @@
+        end
+      end
+
+      context "when first_name missing" do
```
```suggestion
      context "when first_name is missing" do
```

leenagupte commented on spec/services/json_validator_spec.rb:
```diff
@@ -0,0 +1,517 @@
+      let(:data) do
+        valid_data.tap do |valid_data|
+          valid_data[:name].delete(missing)
+        end
```
I think it would be easier to read if this happened in each test, rather than having to scroll up to read it each time.

leenagupte commented on spec/services/json_validator_spec.rb:
```diff
@@ -0,0 +1,517 @@
+        end
+      end
+
+      context "when middle_name missing" do
```
```suggestion
      context "when middle_name is missing" do
```

leenagupte replied:
```suggestion
      context "when middle_name is missing" do
```

leenagupte commented on spec/services/json_validator_spec.rb:
```diff
@@ -0,0 +1,517 @@
+      context "when middle_name missing" do
+        let(:missing) { :middle_name }
+
+        it "returns no error" do
```
```suggestion
        it "returns no errors" do
```

leenagupte replied:
```suggestion
        it "returns no errors" do
```

leenagupte commented on spec/services/json_validator_spec.rb:
```diff
@@ -0,0 +1,517 @@
+    end
+
+    context "when live_in_england in valid" do
+      let(:data) { valid_data.merge(live_in_england: "invalid") }
```
```suggestion
      let(:invalid_data) { valid_data.merge(live_in_england: "invalid") }
```

I think "data" is an overloaded term as it's also something that is returned in `errors`

reggieb replied:
I've renamed it `json_data`. I was tempted to call it `json` but as it's actually a Hash, added the `_data` suffix.

leenagupte commented on spec/services/json_validator_spec.rb:
```diff
@@ -0,0 +1,517 @@
+
+      it "returns a missing keys error" do
+        expect(error["missing_keys"]).to contain_exactly("live_in_england")
+        expect(error["data"]).to be_nil
```
Is there a test for when `error["data"]` is not `nil`?

reggieb replied:
Yes - there are a number of tests that look for `error["data"]` content. I added this one specially because the data content returned by JSONSchemer's `validate` method in this condition contains all the data submitted - so it important that it doesn't get logged. I'll split that out into a separate spec to make that clearer.

leenagupte commented on spec/services/json_validator_spec.rb:
```diff
@@ -0,0 +1,517 @@
+        end
+      end
+
+      context "when last_name missing" do
```
```suggestion
      context "when last_name is missing" do
```

leenagupte replied:
```suggestion
      context "when last_name is missing" do
```

leenagupte commented on app/services/json_validator.rb:
```diff
@@ -0,0 +1,38 @@
+
+  def errors
+    error_details.map do |error|
+      if error["details"].present?
+        error["details"]
+      elsif error["data_pointer"].present?
+        error.slice("data", "data_pointer", "schema_pointer")
+      else
+        {
+          type: "Unknown error",
+          data_keys: data.keys,
+          error_keys: errors.keys,
+        }
+      end
```
The tests check that error["missing_keys"] has the correct value, but there's no reference to "missing_keys" in the code. Where is it coming from?

reggieb replied:
It comes from the JSONSchemer error response. `error["details"]` contains a hash `{ "missing_keys" => ["name of key"] }` when a key is missing.

To be honest, the error behaviour of the gem isn't well documented and has inconsistencies. For example when "details" are present, "data" contains most if not all of the submitted data. Whereas, in other errors, "data" holds a very precise output detailing which data key has failed. So in one condition "data" output needs to be returned in the other it needs to be excluded. 

That's part of the reason I thought a comprehensive set of specs were needed to make sure the error behaviours were covered. I took the specs a little further than the originals testing more variations of the pattern matching to ensure all likely error behaviours were tested.

It's also why I left a catch all in the `else` block as there may be (though I think not) an error condition that I've missed.

leenagupte commented on spec/services/json_validator_spec.rb:
```diff
@@ -0,0 +1,517 @@
+      let(:data) { valid_data.merge(live_in_england: "invalid") }
+
+      it "returns a yes_no definition error against field" do
+        expect(error["data"]).to eq("invalid")
```
Is `error["data"]` either `nil` or `"invalid"`?

reggieb replied:
`error["data"] contains the string passed in that is causing the failure. So here, the string "invalid".

It's an idiom I've seen a lot, that when you need a string that behaves a certain way you use the description of that behaviour as the string. So for an unknown string use "unknown" for example. However, as that seems to be confusing here, I'll use "foo" as used in the old specs.

leenagupte commented on spec/services/json_validator_spec.rb:
```diff
@@ -0,0 +1,517 @@
+      end
+    end
+
+    context "when date_of_birth missing" do
```
```suggestion
    context "when date_of_birth is missing" do
```

leenagupte commented on spec/services/json_validator_spec.rb:
```diff
@@ -0,0 +1,517 @@
+      end
+    end
+
+    context "within date_of_birth" do
```
```suggestion
    context "date_of_birth" do
```

leenagupte commented on spec/services/json_validator_spec.rb:
```diff
@@ -0,0 +1,517 @@
+          valid_data[:date_of_birth].delete(missing) if missing
+          valid_data[:date_of_birth].merge! replace if replace
+        end
+      end
```
This is very confusing to read. I think this set up should be added to each individual test so it is clear exactly what is being tested.

leenagupte commented on spec/services/json_validator_spec.rb:
```diff
@@ -0,0 +1,517 @@
+        end
+      end
+
+      context "when year missing" do
```
```suggestion
      context "when year is missing" do
```

leenagupte commented on spec/services/json_validator_spec.rb:
```diff
@@ -0,0 +1,517 @@
+        end
+      end
+
+      context "when year not a number" do
```
```suggestion
      context "when year is not a number" do
```

leenagupte commented on spec/services/json_validator_spec.rb:
```diff
@@ -0,0 +1,517 @@
+
+      context "when year not a number" do
+        let(:missing) { nil }
+        let(:replace) { { year: "Two Thousand" } }
```
I've forgotten what "replace" is.

leenagupte commented on spec/services/json_validator_spec.rb:
```diff
@@ -0,0 +1,517 @@
+        end
+      end
+
+      context "when month missing" do
```
```suggestion
      context "when month is missing" do
```

leenagupte reviewed (commented):
Thanks for adding the schema validation tests back, as well as adding why the json-schemer gem is being used to the commit message.

About the schema validation tests. Whilst it's great that they are back, they have become very difficult to read. I've gotten about half-way through reviewing the tests, but the comments I've added to the first set of tests apply to the whole:

The set up for the tests is in before blocks or "lets" that are quite far from the actual tests, so by the time you come to read the test, you've forgotten what they're testing. 

I think the tests could be simplified quite a lot if rather than having nested contexts, there was only one context/describe for each field, and each test it contained did the all of the "Arrange, Act and Assert" it needs. This would have the added bonus of making them easy to read.

Lastly, the "describe", "context" and "it" blocks together should read as a sentence. I've corrected the first set of tests to do this, but it applies to all of the others in the file. 

I think I've added inline comments about it, but I think that the `errors` method in the JsonValidator could be clearer. For example, we test the contents of "missing_keys" which isn't in the code, but I don't think we clearly test `error["details"]`, which is. I think it's worth modelling the errors that we receive back from the validator so that we're only passing back useful information to the caller.

Unfortunately I'm not around to talk about these changes in person. I'm happy for the changes in the first commit to go out, so feel free to separate them out, as they feel separate from the json validation changes.

reggieb commented:
@leenagupte thank you for your review.
- I've updated the texts to be more readable as you suggested.
- I've also refactored the specs so that they each contain a more complete construction of the json data
- I've rename `data` to `json_data` so that it isn't confused with the "data" key returned from JSON::Validator.
- I've added a comment about how the errors returned from JSON::Validator behave as it isn't obvious and has taken me a little time to work out.

I've kept the `context`, `let`, `it` structure as that's [good practice for RSpec specs](http://www.betterspecs.org/), but I hope that by bringing the `json_data` construction closer to each `it` clause I've addresses your issue with having to skip back and forth through the spec to work out what's happening.

steveholmes commented:
Nice one! I believe all of Leena's comment are addressed in the updates. (I had earlier approved an earlier version so not able to approve twice :) But :+1: from me) 
Needs a rebase as there are conflicts but otherwise looks good to go.
